pageserver/src/waldecoder.rs

```diff
@@ -804,7 +806,13 @@ pub fn decode_wal_record(record: Bytes) -> DecodedWALRecord {
                     let spcnode = buf.get_u32_le();
                     let dbnode = buf.get_u32_le();
                     let relnode = buf.get_u32_le();
-                    //TODO handle this too?
+                    let mut blk = DecodedBkpBlock::new();
+                    blk.forknum = pg_constants::MAIN_FORKNUM;
+                    blk.rnode_spcnode = spcnode;
+                    blk.rnode_dbnode = dbnode;
+                    blk.rnode_relnode = relnode;
+                    blk.will_drop = true;
```

Comment:
Maybe this should be
```
DecodedBkpBlock::new(MAIN_FORKNUM, spcnode, dbnode, relnode, true)
```
or something like that. The strategy of creating a default all-zeroes `DecodedBkpBlock` and then filling in one field at a time leaves a lot of room for bugs, because the type doesn't indicate what fields are required, which states might be invalid, etc.

"Make Illegal States Unrepresentable" is a powerful strategy for preventing bugs.

If there are multiple constructors I'd expect the all-zeroes one to be `Default::default` or just `fn zeroed() -> Self`, and the others get descriptive names like `from_reltag(&RelTag) -> Self`, etc. If the all-zeroes case is invalid, then we shouldn't allow constructing one at all.

Reply:
You really think that DecodedBkpBlock::new(MAIN_FORKNUM, spcnode, dbnode, relnode, true) 
is less error prone???

It uses positional parameters and you need to remember and control that you are passing them in correct order.
If you write instead  DecodedBkpBlock::new(MAIN_FORKNUM, dbnode, spcnode, relnode, true), and compiler will not notice it and you will spent hours of time trying to understand whats going wrong.

I wonder why Rust with all its error preventing accents still do not have keyword arguments?

Frankly speaking I think that all such refactoring can be done later when code is more or less stabilized. Right now when we have multiple branches performing such refactoring in one branch cause a lot of merge conflicts in other branches. Certainly there are cases when refactoring is really needed like in structs serialization. But IMHO this is not a case.

Reply:
> You really think that DecodedBkpBlock::new(MAIN_FORKNUM, spcnode, dbnode, relnode, true)
is less error prone???

Yes.

> It uses positional parameters and you need to remember and control that you are passing them in correct order.
If you write instead DecodedBkpBlock::new(MAIN_FORKNUM, dbnode, spcnode, relnode, true), and compiler will not notice it and you will spent hours of time trying to understand whats going wrong.

I completely agree. There are a few strategies here to fix that:
- We could use newtypes for each one, so they can't get mixed up.
- If several of these always travel together they should be combined into a single struct, so they don't need to be passed separately. It looks like (spcnode, dbnode, relnode) are just subcomponents of a page address, so maybe they should be rolled into a `PageAddress` type.
- If we're only transferring inner values from one struct to another, we should have a conversion function `Foo::from_bar(b: &Bar)`.

> I wonder why Rust with all its error preventing accents still do not have keyword arguments?

In most places people use types instead. Sometimes they instead use a builder pattern or other strategies, to avoid functions with lots of similar-looking arguments.